File: results/0.75/SQLITE3-DB/sqlite3-db-gpt-4o.rs
use std::fs::File;
use std::io::Read;
use std::env;
use nom::{
    bytes::complete::{tag, take},
    number::complete::{be_u16, be_u32, be_u8},
    IResult,
};

#[derive(Debug)]
struct SQLiteHeader {
    magic_header: [u8; 16],
    page_size: u16,
    file_format_write_version: u8,
    file_format_read_version: u8,
    reserved_space: u8,
    max_embedded_payload_frac: u8,
    min_embedded_payload_frac: u8,
    leaf_payload_frac: u8,
    file_change_counter: u32,
    database_size_pages: u32,
    first_freelist_trunk_page: u32,
    total_freelist_pages: u32,
    schema_cookie: u32,
    schema_format_number: u32,
    default_page_cache_size: u32,
    largest_root_btree_page: u32,
    text_encoding: u32,
    user_version: u32,
    incremental_vacuum_mode: u32,
    application_id: u32,
    reserved_for_expansion: [u8; 20],
    version_valid_for: u32,
    sqlite_version_number: u32,
}

fn parse_sqlite_header(input: &[u8]) -> IResult<&[u8], SQLiteHeader> {
    let (input, magic_header) = take(16usize)(input)?;
    let (input, page_size) = be_u16(input)?;
    let (input, file_format_write_version) = be_u8(input)?;
    let (input, file_format_read_version) = be_u8(input)?;
    let (input, reserved_space) = be_u8(input)?;
    let (input, max_embedded_payload_frac) = be_u8(input)?;
    let (input, min_embedded_payload_frac) = be_u8(input)?;
    let (input, leaf_payload_frac) = be_u8(input)?;
    let (input, file_change_counter) = be_u32(input)?;
    let (input, database_size_pages) = be_u32(input)?;
    let (input, first_freelist_trunk_page) = be_u32(input)?;
    let (input, total_freelist_pages) = be_u32(input)?;
    let (input, schema_cookie) = be_u32(input)?;
    let (input, schema_format_number) = be_u32(input)?;
    let (input, default_page_cache_size) = be_u32(input)?;
    let (input, largest_root_btree_page) = be_u32(input)?;
    let (input, text_encoding) = be_u32(input)?;
    let (input, user_version) = be_u32(input)?;
    let (input, incremental_vacuum_mode) = be_u32(input)?;
    let (input, application_id) = be_u32(input)?;
    let (input, reserved_for_expansion) = take(20usize)(input)?;
    let (input, version_valid_for) = be_u32(input)?;
    let (input, sqlite_version_number) = be_u32(input)?;

    Ok((input, SQLiteHeader {
        magic_header: magic_header.try_into().unwrap(),
        page_size,
        file_format_write_version,
        file_format_read_version,
        reserved_space,
        max_embedded_payload_frac,
        min_embedded_payload_frac,
        leaf_payload_frac,
        file_change_counter,
        database_size_pages,
        first_freelist_trunk_page,
        total_freelist_pages,
        schema_cookie,
        schema_format_number,
        default_page_cache_size,
        largest_root_btree_page,
        text_encoding,
        user_version,
        incremental_vacuum_mode,
        application_id,
        reserved_for_expansion: reserved_for_expansion.try_into().unwrap(),
        version_valid_for,
        sqlite_version_number,
    }))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <path to sqlite3 database file>", args[0]);
        std::process::exit(1);
    }

    let file_path = &args[1];
    let mut file = File::open(file_path).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_sqlite_header(&buffer) {
        Ok((_, header)) => println!("{:#?}", header),
        Err(e) => eprintln!("Failed to parse SQLite header: {:?}", e),
    }
}

File: results/0.75/SQLITE3-DB/sqlite3-db-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

HParser *create_sqlite3_parser() {
    HParser *header_string = h_sequence(
        h_token((const uint8_t *)"SQLite format 3\0", 16),
        NULL
    );
    HParser *page_size = h_be_u16(); // Big-endian 0x10 – 0x11
    HParser *write_version = h_u8();
    HParser *read_version = h_u8();
    HParser *reserved_space = h_u8();
    HParser *max_payload_fraction = h_u8();
    HParser *min_payload_fraction = h_u8();
    HParser *leaf_payload_fraction = h_u8();
    HParser *file_change_counter = h_be_u32(); // Big-endian 0x18 – 0x1B
    HParser *database_size_pages = h_be_u32(); // Big-endian 0x1C – 0x1F
    HParser *first_freelist_trunk_page = h_be_u32(); // Big-endian 0x20 – 0x23
    HParser *total_freelist_pages = h_be_u32(); // Big-endian 0x24 – 0x27
    HParser *schema_cookie = h_be_u32(); // Big-endian 0x28 – 0x2B
    HParser *schema_format_number = h_be_u16(); // Big-endian 0x2C – 0x2D
    HParser *default_cache_size = h_be_u32(); // Big-endian 0x2E – 0x31
    HParser *largest_root_btree_page = h_be_u32(); // Big-endian 0x32 – 0x35
    HParser *text_encoding = h_be_u16(); // Big-endian 0x36 – 0x37
    HParser *user_version = h_be_u32(); // Big-endian 0x38 – 0x3B
    HParser *incremental_vacuum_mode = h_be_u32(); // Big-endian 0x3C – 0x3F
    HParser *application_id = h_be_u32(); // Big-endian 0x40 – 0x43
    HParser *reserved_expansion = h_repeat_n(h_u8(), 12); // 0x44 – 0x4F
    HParser *version_valid_for = h_be_u32(); // Big-endian 0x50 – 0x53
    HParser *sqlite_version = h_be_u32(); // Big-endian 0x54 – 0x57

    return h_sequence(
        header_string,
        page_size,
        write_version,
        read_version,
        reserved_space,
        max_payload_fraction,
        min_payload_fraction,
        leaf_payload_fraction,
        file_change_counter,
        database_size_pages,
        first_freelist_trunk_page,
        total_freelist_pages,
        schema_cookie,
        schema_format_number,
        default_cache_size,
        largest_root_btree_page,
        text_encoding,
        user_version,
        incremental_vacuum_mode,
        application_id,
        reserved_expansion,
        version_valid_for,
        sqlite_version,
        NULL
    );
}

void parse_sqlite3_file(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Failed to open file");
        exit(EXIT_FAILURE);
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    unsigned char *buffer = (unsigned char *)malloc(file_size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        exit(EXIT_FAILURE);
    }

    fread(buffer, 1, file_size, file);
    fclose(file);

    HParser *parser = create_sqlite3_parser();
    HParseResult *result = h_parse(parser, buffer, file_size);

    if (result) {
        printf("SQLite file parsed successfully.\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse SQLite file.\n");
    }

    free(buffer);
    h_parser_destroy(parser);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <sqlite3_db_file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    parse_sqlite3_file(argv[1]);

    return EXIT_SUCCESS;
}